File: cxx/hanoi.cpp
#include <iostream>
#include <cstring>


void move(int num, char from, char mid, char to){
  if(num==1){
    std::cout<<"move "<<from<<"->"<<to<<"\n";
    return;
  }
  move(num-1, from, to, mid);
  move(1, from, mid, to);
  move(num-1, mid, from, to);
}


int main(int argc, char** argv){
  if(argc<2){
    std::cout<<"usage: hanoi <number>\n";
    return -1;
  }
  int n = atoi(argv[1]);
  move(n, 'A', 'B', 'C');
  return 0;
}

File: fortran/hanoi.f90
program hanoi
  implicit none
  integer :: n, i
  character*30 :: char
  n = iargc()
  if (n.EQ.1) then
    call getarg(1, char)
    read(char, "(i30)") i
    call move(i, 'A', 'B', 'C')
  else
    print *, "usage hanoi <number>"
  end if
end program

recursive subroutine move(n, from, mid, to)
  integer :: n
  character :: from, mid, to
  if (n.EQ.1) then
    print *, from, "->", to
  else
    call move(n-1, from, to, mid)
    call move(1, from, mid, to)
    call move(n-1, mid, from, to)
  end if
end subroutine

